app/components/ui/BookmarkModal/BookmarkModal.tsx

```diff
@@ -0,0 +1,216 @@
+
+import styles from "./BookmarkModal.module.scss";
+
+const { siteUrl } = whitelabel;
+
+export const BookmarkModal = ({
+  isOpen,
+  setIsOpen,
+  hit,
+}: {
+  isOpen: boolean;
+  setIsOpen: (val: boolean) => void;
+  hit: SearchHit;
+}) => {
+  // TODO update any types as needed when API work is done
+  interface Folder {
+    id: number;
+    name: string;
+  }
+
+  // TODO: If bookmark already exists, use bookmark name; otherwise,
+  // default to service/resource name. This should be updated with the correct
+  // bookmark name prop when the API is operational
+  const initialName = hit.name;
+  // TODO: Once the service/resource can have an associated bookmark model, we can check
+  // if the bookmark already exists and update various UI on the modal accordingly
+  const newBookmark = false;
+
+  const [folderOptions, setFolderOptions] = useState<Folder[]>([]);
+  const [selectedFolder, setSelectedFolder] = useState<Folder | null>(null);
+  const [newFolder, setNewFolder] = useState("");
+  const [expandFolders, setExpandFolders] = useState(false);
+  const [bookmarkName, setBookmarkName] = useState(initialName);
+  const foldersContainerRef = useRef(null);
+
+  useEffect(() => {
+    // TODO: fetch bookmark folders when the API endpoint is ready. The setTimeout function
+    // is a mock request for the purposes of development.
+    const mockResponse = new Promise((resolve) => {
+      setTimeout(() => {
+        const mockFolders = [
+          { id: 1, name: "Cory's Service List", bookmarks: [] },
+          { id: 2, name: "Fred's Service List", bookmarks: [] },
+          { id: 3, name: "Fredd's Service List", bookmarks: [] },
+          { id: 4, name: "Frewwd's Service List", bookmarks: [] },
+          { id: 5, name: "Frqed's Service List", bookmarks: [] },
+          { id: 8, name: "Frrred's Service List", bookmarks: [] },
+          { id: 9, name: "Fre333d's Service List", bookmarks: [] },
+        ];
+        resolve(mockFolders);
+      }, 500);
+    });
+
+    mockResponse.then((value) => {
+      const response = value as Folder[];
+      setFolderOptions(response);
+    });
+  }, []);
+
+  const handleDropwodnBlur = (
+    event: React.FocusEvent<HTMLDivElement, Element>
+  ) => {
+    // Closes the dropdown unless the newly focused element is a child of the dropdown
+    const newFocus = event.relatedTarget;
+    const folderElement = foldersContainerRef.current as unknown as Element;
+    if (newFocus && !folderElement.contains(newFocus)) {
+      setExpandFolders(false);
+    }
+  };
+
+  const createBookmark = () => {
+    // TODO: when bookmarks/folder API is done, fix this up
+    if (!selectedFolder || !bookmarkName) return;
+    const isNewFolder = selectedFolder.id === -1;
+    if (isNewFolder) {
+      // Create new folder then create new bookmark
+    } else {
+      // Just create new bookmark: pass service/resource ID and folder ID, etc.
+    }
+  };
+
+  const createNewFolder = (name: string) => {
```

Comment:
This name feels a bit misleading, since I assume you don't intend to actually perform an API request to create the folder. I can't actually think of a good name for this, since the best I can come up with is `setNewFolderName()`, but that's basically what `setNewFolder()` already does.

I'm wondering if it would actually be clearer to just have these be separate functions instead of grouping them into a single function. There's really two separate operations happening here: 1) setting the name of the new folder, which is really just for synchronizing with the text input field, and 2) actually selecting a folder for the dropdown menu. It's admittedly a bit weird that the design essentially has two input fields nested within each other, one for the dropdown menu, and one for the text field for naming a new folder, but I think it might be clearer to separate the two actions.

I'll make another comment lower in the file, where this is being called, to comment more specifically on the events we're hooking into.

Reply:
That makes more sense, thanks!